### PR TITLE
Added MIDI Device Selection in Sound Settings

### PR DESCRIPTION
**Pull request by sithmartyr**
Added the ability to select a different MIDI output device, so that you can use a Roland SoundCanvas.

https://youtu.be/zA4txQuixS0

image

### CODE REVIEW

**Comment by mlauss** [comment]
Nice work, works fine in Linux as well:
Screenshot_20221229_144254

**Comment by sithmartyr** [comment]
> Nice work, works fine in Linux as well: Screenshot_20221229_144254

Excellent! I was hoping to test it in Linux later today, but I'm glad you beat me to it lol. 
Thanks for testing that! =)

**Comment by spacekomet** [comment]
Hello. 

Should this suppose to change midi port on the fly during the gameplay? 
Because for me, when I change the midi port in the Sound menu, I must then completely exit the game and run it again to have music from the selected port.  Changing the midi port during the gameplay makes no difference.

**Comment by sithmartyr** [comment]
> Hello.
> 
> Nice idea.
> 
> However, I'm curious. Should this suppose to change midi port on the fly during the gameplay? Because for me, when I change the midi port in the Sound menu, I must then completely exit the game and run it again to have music from the selected port. Changing the midi port during the gameplay makes no difference.
> 
> PS: also using TFE with Linux.

Hello!

I actually tried to make it change on the fly without a restart, but I couldn't figure out how to 'destroy' and re-initiate the MIDI player using the new device without TFE crashing on me.  So, yes, setting the MIDI device and then having to restart is the normal behavior for now.

**Comment by mlauss** [comment]
> > Hello.
> > Nice idea.
> > However, I'm curious. Should this suppose to change midi port on the fly during the gameplay? Because for me, when I change the midi port in the Sound menu, I must then completely exit the game and run it again to have music from the selected port. Changing the midi port during the gameplay makes no difference.
> > PS: also using TFE with Linux.
> 
> Hello!
> 
> I actually tried to make it change on the fly without a restart, but I couldn't figure out how to 'destroy' and re-initiate the MIDI player using the new device without TFE crashing on me. So, yes, setting the MIDI device and then having to restart is the normal behavior for now.

Can you add a note below the drop down box when the midi device has changed, like "restart of TFE required to take effect"  ?

**Comment by sithmartyr** [comment]
> > > Hello.
> > > Nice idea.
> > > However, I'm curious. Should this suppose to change midi port on the fly during the gameplay? Because for me, when I change the midi port in the Sound menu, I must then completely exit the game and run it again to have music from the selected port. Changing the midi port during the gameplay makes no difference.
> > > PS: also using TFE with Linux.
> > 
> > 
> > Hello!
> > I actually tried to make it change on the fly without a restart, but I couldn't figure out how to 'destroy' and re-initiate the MIDI player using the new device without TFE crashing on me. So, yes, setting the MIDI device and then having to restart is the normal behavior for now.
> 
> Can you add a note below the drop down box when the midi device has changed, like "restart of TFE required to take effect" ?

image
Sure can!
Done =)

File: TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp [hunk 1]
```diff
@@ -2231,6 +2234,35 @@ namespace TFE_FrontEndUI
...
+		
+		ImGui::Dummy(ImVec2(0.0f, 20.0f));
+
+		ImGui::LabelText("##ConfigLabel", "MIDI Device");
+		ImGui::LabelText("##ConfigLabel", "# of devices: %i", TFE_MidiDevice::getDeviceCount());
+		ImGui::SetNextItemWidth(512.0f);
+		static int MIDI_CurrentDevIndex = TFE_Settings::getSoundSettings()->midiDevice;
+
+
+		if (ImGui::Combo("##MIDI Device", &MIDI_CurrentDevIndex, MIDI_Devices, IM_ARRAYSIZE(MIDI_Devices))) {
```
Comment (mlauss) [inline TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp:2255]: Why not use "MIDI_DeviceCount" instead of the clunky IM_ARRAYSIZE ?

Comment (sithmartyr) [reply TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp:2255]: I went ahead and changed that.
I had just referenced the other ImGui Combo boxes, but yes using MIDI_DeviceCount makes much more sense, so I went ahead and changed that.
Thanks for pointing that out =)

**Comment by mlauss** [comment]
I like it.

I'm curious though why changing the midi output device on the fly doesn't work.
If you change the device index:
- stop the midi player
- midiDevice::s_midiOut->closePort()  // or integrate that into setDevice() as  _if (s_midiOut->isPortOpen()) s_midiOut->closePort())_
- midiDevice::setDevice(new index)
- restart midi player

Care to try?

**Comment by sithmartyr** [comment]
> I like it.
> 
> I'm curious though why changing the midi output device on the fly doesn't work. If you change the device index:
> 
> * stop the midi player somehow
> * midiDevice::s_midiOut->closePort()  // or integrate that into setDevice() as  _if (s_midiOut->isPortOpen()) s_midiOut->closePort())_
> * midiDevice::setDevice(new index)
> * restart midi player
> 
> Care to try?

I made a change and it does work, _however_, the MIDI instruments do get a little jacked up when you switch the device.
Now you only have to exit the game (not the engine completely) and just click Start again in the main menu and the instruments work fine, so I'll be digging into that a little to see if I can make the instruments work normally after switching.

**Comment by mlauss** [comment]
> > I like it.
> > I'm curious though why changing the midi output device on the fly doesn't work. If you change the device index:
> > 
> > * stop the midi player somehow
> > * midiDevice::s_midiOut->closePort()  // or integrate that into setDevice() as  _if (s_midiOut->isPortOpen()) s_midiOut->closePort())_
> > * midiDevice::setDevice(new index)
> > * restart midi player
> > 
> > Care to try?
> 
> I made a change and it does work, _however_, the MIDI instruments do get a little jacked up when you switch the device. Now you only have to exit the game (not the engine completely) and just click Start again in the main menu and the instruments work fine, so I'll be digging into that a little to see if I can make the instruments work normally after switching.

So at least for me, on Linux, this works beautifully [1].  Please add a check if (MIDI_CurrentDevIndex != TFE_Settings::getSoundSettings()->midiDevice)  before you destroy and reinit the midiplayer. no need to do anything if the selection hasn't changed.

[1] http://mlau.at/tfe_midichange_linux.mp4

File: TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp [hunk 2]
```diff
@@ -2231,6 +2234,37 @@ namespace TFE_FrontEndUI
...
+		const char* MIDI_Devices;
+		std::string MIDI_DevicesStr;
+		
+		for (int i = 0; i < MIDI_DeviceCount; i++) {
+			MIDI_DevicesStr += TFE_MidiDevice::getDeviceNameStr(i) + '\0';
+		}
+		MIDI_Devices = &MIDI_DevicesStr[0];
+		
+		ImGui::Dummy(ImVec2(0.0f, 20.0f));
+
```
Comment (mlauss) [inline TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp:2248]: Maybe add an ImGui::Separator() here.

File: TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp [hunk 3]
```diff
@@ -2231,6 +2234,37 @@ namespace TFE_FrontEndUI
...
+		ImGui::LabelText("##ConfigLabel", "# of devices: %i", TFE_MidiDevice::getDeviceCount());
+		ImGui::SetNextItemWidth(512.0f);
+		static int MIDI_CurrentDevIndex = TFE_Settings::getSoundSettings()->midiDevice;
+
+
+		if (ImGui::Combo("##MIDI Device", &MIDI_CurrentDevIndex, MIDI_Devices, MIDI_DeviceCount)) {
+			sound->midiDevice = MIDI_CurrentDevIndex;
+			TFE_MidiPlayer::destroy();
+			TFE_MidiPlayer::init(MIDI_CurrentDevIndex);
+			MIDI_ShowDevChangeAlert = true;
```
Comment (mlauss) [inline TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp:2259]: Could you please wrap this in
if (MIDI_CurrentDevIndex != TFE_Settings::getSoundSettings()->midiDevice) {
[...]
}
No need to do anything or show the warning if the selection hasn't changed.
Maybe even disable the combobox if only one entry is available?

Comment (sithmartyr) [reply TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp]: I went ahead and added this.
Thanks!

File: TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp (hunk 2 above)
Comment (sithmartyr) [reply TheForceEngine/TFE_FrontEndUI/frontEndUi.cpp:2248]: I made a commit that I think makes this look OK without the separator.  I'm trying to keep the overall appearance of the sound config menu consistent.

**Comment by mlauss** [comment]
I've implemented a cache to send the current midi config to the newly selected synthesizer.
Care to take it for a test?
just clone this branch: https://github.com/mlauss/TheForceEngine/tree/test-audio-select
it *should* work on windows, but i've only built it on linux so far.

**Comment by luciusDXL** [comment]
This has been resolved.

**Pull request closed without merging**